util/src/main/java/tc/oc/pgm/util/menu/InventoryMenu.java

```diff
@@ -0,0 +1,369 @@
+        || event.getCurrentItem() == null
+        || event.getCurrentItem().getItemMeta() == null
+        || event.getCurrentItem().getItemMeta().getDisplayName() == null) return;
+
+    if (event.getWhoClicked() instanceof Player) {
+      Player player = ((Player) event.getWhoClicked());
+      if (isViewing(player)) {
+        ItemStack clicked = event.getCurrentItem();
+        for (InventoryMenuItem item : this.inventoryMenuItems) {
+          if (item == null) continue;
+
+          if (clicked.equals(item.createItem(player))) {
+            item.onInventoryClick(this, player, event.getClick());
+          }
+        }
+      }
+    }
+  }
+
+  @EventHandler
+  public void onInventoryClose(final InventoryCloseEvent event) {
+    // Remove viewing of menu upon inventory close
+    Player player = (Player) event.getPlayer();
+    this.remove(player);
+  }
+
+  @EventHandler
+  public void onWorldUnload(WorldUnloadEvent event) {
+    if (this.world != event.getWorld()) return;
+    disableInventory();
+  }
+
+  public void disableInventory() {
+    viewing.keySet().forEach(this::closeWindow);
+    viewing.clear();
+    HandlerList.unregisterAll(this);
+  }
+
+  public void enableInventory() {
+    Bukkit.getPluginManager().registerEvents(this, BukkitUtils.getPlugin());
+  }
+
+  private List<InventoryMenuItem> applyPatternAndAddPages(
+      List<InventoryMenuItem> items, MenuArranger menuArranger, boolean pagesPossible) {
+
+    List<InventoryMenuItem> mutableItems = new ArrayList<>(items);
+    if (!pagesPossible) return items;
+
+    // Quick exit if we dont need any pages
+    if (menuArranger.automatedPaginationLimit() > mutableItems.size())
+      return menuArranger.arrangeItems(mutableItems);
+
+    // We need pages!!
+    rows++;
+
+    List<List<InventoryMenuItem>> pages = new ArrayList<>();
+
+    // Put items into pages
+    List<InventoryMenuItem> page = new ArrayList<>();
+    for (int i = 0; !mutableItems.isEmpty(); i++) {
+      page.add(mutableItems.remove(0));
+      if (i + 1 == menuArranger.automatedPaginationLimit() || mutableItems.isEmpty()) { // new page
+        pages.add(new ArrayList<>(menuArranger.arrangeItems(page)));
+        page.clear();
+        i = 0;
+      }
+    }
+
+    // Insert pagination items on every page
+    for (int i = 0; i < pages.size(); i++) {
+      List<InventoryMenuItem> currentPage = pages.get(i);
+      currentPage.add(null);
+      currentPage.add(null);
+      currentPage.add(
+          // Is there a previous page?
+          i == 0 ? null : new PageInventoryMenuItem(pages.get(i - 1), menuArranger, false));
+      currentPage.add(null);
+      currentPage.add(null);
+      currentPage.add(null);
+      currentPage.add(
+          // Is there a next page?
+          pages.size() - 1 == i
+              ? null
+              : new PageInventoryMenuItem(pages.get(i + 1), menuArranger, true));
+      currentPage.add(null);
+      currentPage.add(null);
```

Comment:
```suggestion
      List<InventoryMenuItem> currentPage = pages.get(i);
      for (int item = 0; item < ROW_WIDTH; item++) currentPage.add(null);
      if (i > 0) currentPage.set(2, new PageInventoryMenuItem(pages.get(i - 1), menuArranger, false));
      if (i < pages.size() - 1) currentPage.set(6, new PageInventoryMenuItem(pages.get(i - 1), menuArranger, false));
```

You can simply add 9 items on a loop and then set the 2 overrides for previous/next pages, adding so many currentPage.add(null) is a bit dirty